app/assets/javascripts/common/admins.js

```diff
@@ -0,0 +1,213 @@
+    this.toggleAccessTreeVisibility(target.value === ACCESS_LEVEL_POWER_USER)
+  },
+
+  toggleItemCollapsed: function (element) {
+    const collapsed = element.classList.contains("collapsed")
+
+    if (collapsed) {
+      element.classList.remove("collapsed")
+    } else {
+      element.classList.add("collapsed")
+    }
+  },
+
+  onFacilityAccessItemToggled: function ({ target }) {
+    const children = Array.from(target.closest("li").childNodes)
+    const parentItem = target.closest(".access-item")
+    const wrapper = children.find(containsClass("access-item-wrapper"))
+    if (wrapper) {
+      this.toggleItemCollapsed(parentItem)
+    }
+  },
+
+  updateParentCheckedState: function (element, selector) {
+    // find parent and sibling checkboxes
+    const parent = (element.closest(["ul"]).parentNode).querySelector(selector)
+    const siblings = nodeListToArray(selector, parent.closest("li").querySelector(["ul"]))
+
+    // get checked state of siblings
+    // are every or some siblings checked (using Boolean as test function)
+    const checkStatus = siblings.map(check => check.checked)
+    const every = checkStatus.every(Boolean)
+    const some = checkStatus.some(Boolean)
+
+    // check parent if all siblings are checked
+    // set indeterminate if not all and not none are checked
+    parent.checked = every
+    parent.indeterminate = some && !every
+
+    // recurse until check is the top most parent
+    if (element !== parent) this.updateParentCheckedState(parent, selector)
+  },
+
+  updateChildrenCheckedState: function (parent, selector) {
+    // check/uncheck children (includes check itself)
+    const children = nodeListToArray(selector, parent.closest("li"))
+
+    children.forEach(child => {
+      // reset indeterminate state for children
+      child.indeterminate = false
+      child.checked = parent.checked
+    })
+  },
+
+  onAsyncLoaded: function () {
+    const _self = this
+    document.addEventListener('render_async_load', function (_event) {
+      _self.resourceRowCollapseListener()
+    });
+  }
+}
+AdminAccessInvite = function () { }
+
+AdminAccessInvite.prototype = Object.create(AdminAccess.prototype)
+
+AdminAccessInvite.prototype = Object.assign(AdminAccessInvite.prototype, {
+  selectAllFacilitiesInput: () => document.getElementById("select-all-facilities-input"),
+
+  updateIndeterminateCheckboxes: function () {
+    // list of all checkboxes under facilityAccess
+    const checkboxes = nodeListToArray(ACCESS_LIST_INPUT_SELECTOR, this.facilityAccess())
+
+    // go through all the checkboxes that are pre-checked and update their parents accordingly
+    for (const checkbox of checkboxes) {
+      if (!checkbox.checked) continue
+
+      // a large tree can take a lot of time to load on the DOM,
+      // so we queue up our updates by requesting frames so as to not cause overwhelming repaints
+      const _self = this
+      requestAnimationFrame(function () {
```

Comment:
2/2:

This is the second useful change in this file, when we're in the edit screen, and we pre-check the appropriate check-boxes, we still need to update the parents for indeterminate state ([can't provide indeterminate state from HTML](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#Indeterminate_state_checkboxes)).

This needs to loop through all the checked checkboxes and apply `updateParentCheckedState` on each, which can be quite expensive and I've seen slow rendering times because of it.

Here we queue up these updates so that it doesn't block the UI thread and make the page slow to render. `requestAnimationFrame` works well here.